src/array/cpu/spmat_op_impl_coo.cc

```diff
@@ -266,29 +266,57 @@ CSRMatrix COOToCSR(COOMatrix coo) {
   const IdType* row_data = static_cast<IdType*>(coo.row->data);
   const IdType* col_data = static_cast<IdType*>(coo.col->data);
   const IdType* data = COOHasData(coo)? static_cast<IdType*>(coo.data->data) : nullptr;
+
   NDArray ret_indptr = NDArray::Empty({N + 1}, coo.row->dtype, coo.row->ctx);
   NDArray ret_indices;
   NDArray ret_data;
 
-  IdType* Bp = static_cast<IdType*>(ret_indptr->data);
-
-  std::fill(Bp, Bp + N, 0);
-  for (int64_t i = 0; i < NNZ; ++i) {
-    Bp[row_data[i]]++;
+  bool row_sorted = coo.row_sorted;
+  bool col_sorted = coo.col_sorted;
+  if (!row_sorted) {
+    // It is possible that the flag is simply not set (default value is false),
+    // so we still perform a linear scan to check the flag.
+    std::tie(row_sorted, col_sorted) = COOIsSorted(coo);
   }
 
-  // cumsum
-  for (int64_t i = 0, cumsum = 0; i < N; ++i) {
-    const IdType temp = Bp[i];
-    Bp[i] = cumsum;
-    cumsum += temp;
-  }
-  Bp[N] = NNZ;
+  if (row_sorted) {
```

Comment:
i don't understand here. if col_sorted is true, we also don't need to go through the complex computation in the else branch. it seems to me that this case isn't optimized.

Reply:
I c. One thing I didn't make it clear is that the `col_sorted` flag only takes effect when the `row_sorted` is true. The `col_sorted` means each column **per row** is also sorted. Should I change another name?